presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/execution/thrift/ThriftSqlFunctionExecutor.java

```diff
@@ -74,26 +79,68 @@ public void setBlockEncodingSerde(BlockEncodingSerde blockEncodingSerde)
+        ThriftFunctionHandle thriftFunctionHandle = new ThriftFunctionHandle(
+                functionId.getFunctionName().toString(),
+                functionId.getArgumentTypes().stream()
+                        .map(TypeSignature::toString)
+                        .collect(toImmutableList()),
+                returnType.toString(),
+                functionHandle.getVersion());
+        ThriftUdfService thriftUdfService = thriftUdfClient.get(Optional.of(functionImplementation.getLanguage().getLanguage()));
+        return invokeUdfAndHandleException(thriftUdfService, thriftFunctionHandle, page)
+                .thenApply(result -> getResultBlock(result, returnType));
+    }
+
+    public static CompletableFuture<ThriftUdfResult> invokeUdfAndHandleException(
+            ThriftUdfService thriftUdfService,
+            ThriftFunctionHandle functionHandle,
+            ThriftUdfPage page)
+    {
+        ListenableFuture<ThriftUdfResult> resultFuture;
         try {
-            return toCompletableFuture(thriftUdfClient.get(Optional.of(functionImplementation.getLanguage().getLanguage())).invokeUdf(
-                    new ThriftFunctionHandle(
-                            functionId.getFunctionName().toString(),
-                            functionId.getArgumentTypes().stream()
-                                    .map(TypeSignature::toString)
-                                    .collect(toImmutableList()),
-                            returnType.toString(),
-                            functionHandle.getVersion()),
-                    page))
-                    .thenApply(result -> getResultBlock(result, returnType));
+            resultFuture = thriftUdfService.invokeUdf(functionHandle, page);
         }
         catch (ThriftUdfServiceException | TException e) {
-            throw new PrestoException(GENERIC_INTERNAL_ERROR, e);
+            throw new RuntimeException(e);
         }
+
+        CompletableFuture<ThriftUdfResult> future = new CompletableFuture<>();
+        future.exceptionally(throwable -> {
+            if (throwable instanceof CancellationException) {
+                resultFuture.cancel(true);
+            }
+            return null;
+        });
+
+        FutureCallback<ThriftUdfResult> callback = new FutureCallback<ThriftUdfResult>()
+        {
+            @Override
+            public void onSuccess(ThriftUdfResult result)
+            {
+                future.complete(result);
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+                PrestoException prestoException = t instanceof ThriftUdfServiceException ?
+                        toPrestoException((ThriftUdfServiceException) t) :
+                        new PrestoException(GENERIC_INTERNAL_ERROR, t);
+                future.completeExceptionally(prestoException);
+            }
+        };
+        addCallback(resultFuture, callback, directExecutor());
+        return future;
```

Comment:
@rongrong FYI, after some debugging, I found that original way of catching the ThriftUdfServiceException was incorrect.

The signature of `thriftUdfService.invokeUdf` says that it throws `ThriftUdfServiceException`, but since it is returning a ListenableFuture, the exception is never thrown at the function call. Instead, the exception is only thrown when RemoteProjectOperator reads its output. Hence, we need some fancy Future mangling to make the exception catching-and-rethrowing work.